pq-crypto/kyber_r3/cbd.c

```diff
@@ -0,0 +1,104 @@
+* Returns 32-bit unsigned integer loaded from x
+**************************************************/
+static uint32_t load32_littleendian(const uint8_t x[4]) {
+    uint32_t r;
+    r  = (uint32_t)x[0];
+    r |= (uint32_t)x[1] << 8;
+    r |= (uint32_t)x[2] << 16;
+    r |= (uint32_t)x[3] << 24;
+    return r;
+}
+
+/*************************************************
+* Name:        load24_littleendian
+*
+* Description: load 3 bytes into a 32-bit integer
+*              in little-endian order
+*              This function is only needed for Kyber-512
+*
+* Arguments:   - const uint8_t *x: pointer to input byte array
+*
+* Returns 32-bit unsigned integer loaded from x (most significant byte is zero)
+**************************************************/
+static uint32_t load24_littleendian(const uint8_t x[3]) {
+    uint32_t r;
+    r  = (uint32_t)x[0];
+    r |= (uint32_t)x[1] << 8;
+    r |= (uint32_t)x[2] << 16;
+    return r;
+}
+
+
+/*************************************************
+* Name:        cbd2
+*
+* Description: Given an array of uniformly random bytes, compute
+*              polynomial with coefficients distributed according to
+*              a centered binomial distribution with parameter eta=2
+*
+* Arguments:   - poly *r:            pointer to output polynomial
+*              - const uint8_t *buf: pointer to input byte array
+**************************************************/
+static void cbd2(poly *r, const uint8_t buf[2 * KYBER_N / 4]) {
+    unsigned int i, j;
+
+    for (i = 0; i < KYBER_N / 8; i++) {
+        uint32_t t  = load32_littleendian(buf + 4 * i);
+        uint32_t d  = t & 0x55555555;
+        d += (t >> 1) & 0x55555555;
+
+        for (j = 0; j < 8; j++) {
+            int16_t a = (d >> (4 * j + 0)) & 0x3;
+            int16_t b = (d >> (4 * j + 2)) & 0x3;
+            r->coeffs[8 * i + j] = a - b;
+        }
+    }
+}
+
+/*************************************************
+* Name:        cbd3
+*
+* Description: Given an array of uniformly random bytes, compute
+*              polynomial with coefficients distributed according to
+*              a centered binomial distribution with parameter eta=3
+*              This function is only needed for Kyber-512
+*
+* Arguments:   - poly *r:            pointer to output polynomial
+*              - const uint8_t *buf: pointer to input byte array
+**************************************************/
+static void cbd3(poly *r, const uint8_t buf[3 * KYBER_N / 4]) {
+    unsigned int i, j;
+
+    for (i = 0; i < KYBER_N / 4; i++) {
+        uint32_t t  = load24_littleendian(buf + 3 * i);
+        uint32_t d  = t & 0x00249249;
+        d += (t >> 1) & 0x00249249;
+        d += (t >> 2) & 0x00249249;
+
+        for (j = 0; j < 4; j++) {
+            int16_t a = (d >> (6 * j + 0)) & 0x7;
+            int16_t b = (d >> (6 * j + 3)) & 0x7;
+            r->coeffs[4 * i + j] = a - b;
+        }
+    }
+}
+
+void cbd_eta1(poly *r, const uint8_t buf[KYBER_ETA1 * KYBER_N / 4]) {
```

Comment:
If s2n is compiled as a static library this symbol will potentially collide with any other functions with the same name. I know this is an import, but it's usually best practice to prefix all of your functions with the library name to reduce potential collisions (in this case it would be `s2n`). Assuming this code eventually ends up in aws-lc, we will definitely have to deal with the duplication for some period of time.

In the case of s2n as a shared object, we hide internal symbols by default (assuming it's not running any tests).

Reply:
I updated the PR description to be more explicit about this issue. I had defined a macro in `params.h` (`#define S2N_KYBER_512_R3_NAMESPACE(s) s2n_kyber_512_r3_##s`) to maintain symbol/namespace boundaries to avoid collisions, but it definitely wasn't obvious from just looking at the PR originally. 

I also prefixed all the macros appropriately since there is naming overlap with other KEMs.